apollo-runtime/src/main/java/com/apollographql/android/RealApolloCall.java

```diff
@@ -0,0 +1,134 @@
+package com.apollographql.android;
+
+import com.apollographql.android.api.graphql.Operation;
+import com.apollographql.android.api.graphql.Response;
+import com.apollographql.android.api.graphql.ResponseFieldMapper;
+import com.apollographql.android.api.graphql.ScalarType;
+import com.squareup.moshi.JsonAdapter;
+import com.squareup.moshi.Moshi;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+import okhttp3.Call;
+import okhttp3.MediaType;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okio.Buffer;
+
+final class RealApolloCall implements ApolloCall {
+  private static final MediaType MEDIA_TYPE = MediaType.parse("application/graphql; charset=utf-8");
+
+  private final Operation operation;
+  private final Moshi moshi;
+  private final Request baseRequest;
+  private final okhttp3.Call.Factory callFactory;
+  private final ResponseBodyConverter responseBodyConverter;
+  private final AtomicReference<Call> callRef = new AtomicReference<>(null);
+
+  RealApolloCall(Operation operation, Moshi moshi, okhttp3.Call.Factory callFactory, Request baseRequest,
+      ResponseFieldMapper responseFieldMapper, Map<ScalarType, CustomTypeAdapter> customTypeAdapters) {
+    this.operation = operation;
+    this.moshi = moshi;
+    this.callFactory = callFactory;
+    this.baseRequest = baseRequest;
+    this.responseBodyConverter = new ResponseBodyConverter(operation, responseFieldMapper, customTypeAdapters);
+  }
+
+  RealApolloCall(Operation operation, Moshi moshi, okhttp3.Call.Factory callFactory, Request baseRequest,
+      ResponseBodyConverter responseBodyConverter) {
+    this.operation = operation;
+    this.moshi = moshi;
+    this.callFactory = callFactory;
+    this.baseRequest = baseRequest;
+    this.responseBodyConverter = responseBodyConverter;
+  }
+
+  @Override public void cancel() {
+    Call httpCall = callRef.get();
+    if (httpCall != null) {
+      httpCall.cancel();
+    }
+  }
+
+  @Override public <T extends Operation.Data> Response<T> execute() throws IOException {
+    RequestBody requestBody = requestBody(operation);
+    Request request = baseRequest.newBuilder()
+        .post(requestBody)
+        .build();
+
+    Call call = callFactory.newCall(request);
+    if (callRef.compareAndSet(null, call)) {
```

Comment:
yes we need this as RealApolloCall keeps ref on okhttp3.Call in order to cancel request.